lib/charms/lte_core_interface/v0/lte_core_interface.py

```diff
@@ -0,0 +1,246 @@
+
+    def __init__(self, handle: Handle, mme_ipv4_address: str):
+        """Init."""
+        super().__init__(handle)
+        self.mme_ipv4_address = mme_ipv4_address
+
+    def snapshot(self) -> dict:
+        """Returns snapshot."""
+        return {"mme_ipv4_address": self.mme_ipv4_address}
+
+    def restore(self, snapshot: dict) -> None:
+        """Restores snapshot."""
+        self.mme_ipv4_address = snapshot["mme_ipv4_address"]
+
+
+class CoreRequirerCharmEvents(CharmEvents):
+    """List of events that the core requirer charm can leverage."""
+
+    lte_core_available = EventSource(CoreAvailableEvent)
+
+
+class CoreRequires(Object):
+    """Class to be instantiated by the charm requiring the core."""
+
+    on = CoreRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+
+    @staticmethod
+    def _relation_data_is_valid(remote_app_relation_data: dict) -> bool:
+        try:
+            validate(instance=remote_app_relation_data, schema=REQUIRER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError as e:
+            logger.error("Invalid relation data: %s", e)
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        relation = event.relation
+        if not relation:
+            logger.warning("No relation with name %s", self.relationship_name)
+            return
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relationship_name)
+            return
+        remote_app_relation_data = relation.data[relation.app]
+        if not self._relation_data_is_valid(dict(remote_app_relation_data)):
+            logger.warning(
+                "Provider relation data did no pass JSON Schema validation: \n%s",
+                event.relation.data[event.app],  # type: ignore[index]
+            )
+            return
+        self.on.lte_core_available.emit(
+            mme_ipv4_address=remote_app_relation_data["mme_ipv4_address"],
+        )
+
+
+class CoreProvides(Object):
+    """Class to be instantiated by the charm providing the core."""
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.relationship_name = relationship_name
+        self.charm = charm
+
+    @staticmethod
+    def mme_ipv4_address_is_valid(mme_ipv4_address: str) -> bool:
```

Comment:
This should probably be a private method (unless we intend on users of this class to directly call it?)

Reply:
Fixed.